### PR TITLE
ZCS-4798 : RecoverAccount API and it's implementation

### PR DESCRIPTION
**Pull request by chinmay15**
**Problem :** RecoverAccount API and it's implementation
**Fix :** Soap API and implementation  created.
RecoverAccountRequest with 2 operations
1. getRecoveryEmail : in response user will get recovery email(masked) address only if it's verified.
2. sendRecoveryCode : 8 character(alphanumeric) code is sent to user's recovery email, only if it's verified.
3. updated soap.txt 

**Testing done :**
1. Manual testing of overall feature.
2. Manual testing of getRecoveryEmail
3. Manual testing of sendRecoveryCode (with and without alias)
4. Added unit tests

### CODE REVIEW

File: store-conf/conf/msgs/ZsMsg.properties [hunk 1]
```diff
@@ -938,3 +938,19 @@ verifyRecoveryEmailBodyHtml =\
   Recovery email verification code: {0}</br>\
   The code expires by {1}\
   </div>
+
+# password reset code verification
+sendPasswordRecoveryEmailSubject = Request to recover account using recovery email address by {0}
```
Comment (shrikant-prasad) [inline store-conf/conf/msgs/ZsMsg.properties]: Following format is mentioned in ticket ZCS-4781. Better to use the same template for the mail.
Subject: Reset your %DOMAIN% password

We have received a password reset request for your account, %EMAIL%.

Temporary Reset Code: X X X X X X X X
Use by: %EXPIRY%

File: store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        String email = req.getEmail();
+        try {
+            InternetAddress ia = new JavaMailInternetAddress(email, true);
+            ia.validate();
+        } catch (AddressException ae) {
+            ZimbraLog.account.debug("%s Invalid email address: %s", LOG_OPERATION, email);
+            throw ServiceException.INVALID_REQUEST("invalid email address: " + email, ae);
+        }
+
+        Account user = Provisioning.getInstance().getAccountByName(email);
```
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java]: user account can be null. Null check should be added.
No need of validating the internet address.

File: store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        if (user.getPrefPasswordRecoveryAddressStatus() == null || !user.getPrefPasswordRecoveryAddressStatus().isVerified()) {
+            ZimbraLog.account.debug("%s Verified recovery email is not found for %s", LOG_OPERATION, email);
+            throw ServiceException.NOT_FOUND("Verified recovery email is not found");
```
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java]: should not throw Not found exception. Better use Failure exception with message: something is not right, please contact admin.

File: store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        if (StringUtil.isNullOrEmpty(recoveryEmail)) {
+            ZimbraLog.account.debug("%s Recovery email missing or unverified for %s", LOG_OPERATION, email);
+            throw ServiceException.NOT_FOUND("Recovery email address for " + email + " is missing or not verified");
+        }
+
+        RecoverAccountResponse resp = new RecoverAccountResponse();
+        switch (op) {
+            case GET_RECOVERY_EMAIL:
+                recoveryEmail = StringUtil.maskEmail(recoveryEmail);
+                if (!StringUtil.isNullOrEmpty(recoveryEmail)) {
```
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java]: already checked that recoveryEmail is not null above.

File: soap/src/java/com/zimbra/soap/mail/type/PasswordResetOperation.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    @Override
+    public String toString() {
+        return name;
+    }
+
+    public static PasswordResetOperation fromString(String name) {
+        if (name == null) {
+            name = "";
+        }
+        return nameToPasswordRestOperations.get(name);
```
Comment (fatkudu) [inline soap/src/java/com/zimbra/soap/mail/type/PasswordResetOperation.java]: I'd just replace the body with `return nameToPasswordRestOperations.get(Strings.nullToEmpty(name));

File: store-conf/conf/msgs/ZsMsg.properties (hunk 1 above)
Comment (chinmay15) [reply store-conf/conf/msgs/ZsMsg.properties]: done

File: store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java (hunk 2 above)
Comment (chinmay15) [reply store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java]: done

File: store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java (hunk 3 above)
Comment (chinmay15) [reply store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java]: done

File: store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java (hunk 4 above)
Comment (chinmay15) [reply store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java]: done

File: store/src/java/com/zimbra/cs/service/util/JWEUtil.java [hunk 6]
```diff
@@ -19,13 +19,45 @@
...
 
+import com.google.common.collect.Maps;
 import com.zimbra.common.service.ServiceException;
 import com.zimbra.common.util.BlobMetaData;
 import com.zimbra.common.util.BlobMetaDataEncodingException;
 import com.zimbra.cs.account.AuthTokenKey;
 import com.zimbra.cs.account.DataSource;
 
 public class JWEUtil {
+    public enum Constants {
```
[the remark this answers is not in the document]
Comment (chinmay15) [reply store/src/java/com/zimbra/cs/service/util/JWEUtil.java]: done

File: soap/src/java/com/zimbra/soap/mail/type/PasswordResetOperation.java (hunk 5 above)
Comment (chinmay15) [reply soap/src/java/com/zimbra/soap/mail/type/PasswordResetOperation.java]: done

File: store/src/java-test/com/zimbra/cs/service/mail/RecoverAccountTest.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+/*
+ * ***** BEGIN LICENSE BLOCK *****
```
Comment (shrikant-prasad) [inline store/src/java-test/com/zimbra/cs/service/mail/RecoverAccountTest.java]: don't need this class.

Comment (chinmay15) [reply store/src/java-test/com/zimbra/cs/service/mail/RecoverAccountTest.java]: removed. also changed enum as per discussion.

**Comment by chinmay15** [comment]
merged into feature/PasswordReset branch

commit 61a2c9f3f21fae95fb5bfe3956d593943d2fb268
Author: Chinmay Pathak <15.chinmay@gmail.com>
Date:   Tue May 22 18:23:24 2018 +0530

    ZCS-4798 : RecoverAccount API and it's implementation

**Pull request merged**